### ISSUE TITLE
False positive: UseCheckOrError

### ISSUE THREAD

**Issue by BraisGabin** [arturbosch/detekt#2514]
## Expected Behavior

This code should not be flagged

```kotlin
val lambda: (Throwable) -> Unit = {
  if (it is IllegalStateException) {
    println("something")
  } else {
    throw IllegalStateException(it)
  }
}
```

## Observed Behavior
It is flagged

## Context
Replacing `throw IllegalStateException(error)` with `error(error)` changes the behaviour so it should not be flagged.
## Your Environment
* Version of detekt used: 1.6.0

**Comment by schalkms** [comment]
Okay. Have you tried to rewrite the lambda to something like this?

```kotlin
val lambda: (Throwable) -> Unit = {
  check(it is IllegalStateException) { throw IllegalStateException(it) }
  println("something")
}
```

**Comment by BraisGabin** [comment]
My snipped was just a simplification. The point is that `throw IllegalStateException(it)` is not the same as `error(it)` unless it is a `String`. We need type resolution to fix this problem.

**Comment by schalkms** [comment]
@BraisGabin mine was also about simplification. The whole rule is about simplification. You could use either `error` or `check`. I'm curios, what part of my code snippet didn't work for you @BraisGabin ?

**Comment by BraisGabin** [comment]
```
when(throwable) {
    is NumberFormatException -> println("a")
    is IllegalArgumentException -> println("b")
    is NullPointerException -> println("c")
    else -> throw IllegalStateException(throwable)
}
```

The code is something like this but with domain exceptions (users not logged, user disabled, user no verfied...).

**Comment by arturbosch** [comment]
The basic rule part could check for a string literal argument.
The advanced part when BindingContext is set could resolve the type ^^.

**Comment by schalkms** [comment]
Now I see the issue. `error()` and `check()` don't provide an overload for wrapping exceptions. That's unfortunate. To solve this problem, one needs to use the `BindingContext` and check for exactly 1 argument of type `String`.
